File: dart/commection/lib/simpleHttpType/request.dart
import 'package:commection/lazy.dart';
import 'package:commection/simpleHttpType/url.dart';
import 'package:meta/meta.dart';
import 'package:narumincho_json/narumincho_json.dart';

@immutable
sealed class SimpleRequest {
  const SimpleRequest();
}

@immutable
final class SimpleRequestGet implements SimpleRequest {
  const SimpleRequestGet({
    required this.url,
  });

  final SimpleUrl url;
}

@immutable
final class SimpleRequestOption implements SimpleRequest {
  const SimpleRequestOption({
    required this.url,
  });

  final SimpleUrl url;
}

@immutable
final class SimpleRequestPost implements SimpleRequest {
  const SimpleRequestPost({
    required this.url,
    required this.header,
    required this.bodyLazy,
  });

  final SimpleUrl url;
  final SimpleRequestHeader header;
  final Lazy<JsonValue> bodyLazy;
}

@immutable
class SimpleRequestHeader {
  const SimpleRequestHeader({required this.authorizationBearerValue});

  final String? authorizationBearerValue;
}

File: rust/commection/src/simpleHttpType/request.rs
use super::url::SimpleUrl;

pub enum SimpleRequest {
    Get(SimpleRequestGet),
    Option(SimpleRequestOption),
    Post(SimpleRequestPost),
}

pub struct SimpleRequestGet {
    pub url: SimpleUrl,
}

pub struct SimpleRequestOption {
    pub url: SimpleUrl,
}

pub struct SimpleRequestPost {
    pub url: SimpleUrl,
    pub header: SimpleRequestHeader,
    pub body: once_cell::sync::Lazy<serde_json::Value>,
}

pub struct SimpleRequestHeader {
    pub authorization_bearer_value: Option<String>,
}
